lib/core/include/irods/configuration_hook_manager.hpp

```diff
@@ -0,0 +1,28 @@
+#ifndef IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+#define IRODS_CONFIGURATION_HOOK_MANAGER_HPP
+
+/// \file
+
+#include <nlohmann/json.hpp>
+
+#include <vector>
+#include <string>
+
+namespace irods::experimental::configuration_reload
+{
+    /// \brief Hooks are functions called with the pathname of the json value that has changed, as well as the new
+    /// operation performed, removal versus replacement versus insertion etc
+    using hook = std::function<void(const std::string&, const nlohmann::json&)>;
+
+    /// \brief adds a hook to be run whenever a key on the list of hooks changes.
+    /// \param [in] hooks A list of strings to match
+    /// \param [in] hook The function to run when the string matches
+    void add_hook(const auto& hooks, hook&& hook);
```

Comment:
`const auto& hooks` is going to need an example.

What is meant by `"A list of strings to match"`?
We need to be clear on the requirements of this function, else no one will know how to use it properly.

Put yourself in the shoes of a developer who is seeing this for the first time.
What do you think they would think after reading the documentation?

We want to remove as much guesswork as possible.

Reply:
Okay, the idea is that they can pass it something like an array or std::vector, or std::list I guess.

You're definitely right that need to expand on this